src/clj/puppetlabs/services/puppet_profiler/puppet_profiler_core.clj

```diff
@@ -1,24 +1,131 @@
+  (let [count (.getCount timer)
+        mean (metrics/mean-millis timer)]
+    {:metric metric-name
+     :count count
+     :mean mean
+     :aggregate (* count mean)}))
+
+(defn inline-metric
+  [[metric-name timer]]
+  (let [count (.getCount timer)]
+    {:metric metric-name
+     :count count}))
+
+(schema/defn assoc-metrics-data
+  [status :- {schema/Any schema/Any}
+   metrics-profiler :- (schema/maybe MetricsPuppetProfiler)]
+  (if metrics-profiler
+    (let [function-metrics (->> (map function-metric (.getFunctionTimers metrics-profiler))
+                                (sort-by :aggregate)
+                                reverse)
+          resource-metrics (->> (map resource-metric (.getResourceTimers metrics-profiler))
+                                (sort-by :aggregate)
+                                reverse)
+          catalog-metrics (->> (map catalog-metric (.getCatalogTimers metrics-profiler))
+                               (sort-by :aggregate)
+                               reverse)
+          inline-metrics (->> (map inline-metric (.getInliningTimers metrics-profiler))
+                               (sort-by :count)
+                               reverse)]
+      (-> status
+          ;; TODO: decide whether 50 is a reasonable max here
+          (assoc-in [:experimental :function-metrics] (take 50 function-metrics))
+          (assoc-in [:experimental :resource-metrics] (take 50 resource-metrics))
+          (assoc-in [:experimental :catalog-metrics] catalog-metrics)
+          (assoc-in [:experimental :inline-metrics] inline-metrics)))
+    status))
+
+;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;
 ;;; Public
 
-(schema/defn ^:always-validate create-profiler :- Profiler
-  "Given profiler configuration information, create and return a new profiler
-  (or nil if the configuration indicates that no profiler should be used)."
-  [profiler-config :- ProfilerConfig]
-  (if (ks/to-bool (get profiler-config :enabled))
-    (LoggingPuppetProfiler.)))
\ No newline at end of file
+(schema/defn ^:always-validate v1-status :- status-core/StatusCallbackResponse
+  [metrics-profiler :- (schema/maybe MetricsPuppetProfiler)
+   level :- status-core/ServiceStatusDetailLevel]
+  (let [level>= (partial status-core/compare-levels >= level)]
+    {:state :running
+     :status (cond->
+               ;; no status info at ':critical' level
+               {}
+               ;; no extra status at ':info' level yet
+               (level>= :info) identity
+               (level>= :debug) (assoc-metrics-data metrics-profiler))}))
+
+(schema/defn ^:always-validate initialize :- MetricsProfilerServiceContext
+  [config :- MetricsProfilerServiceConfig
+   hostname :- schema/Str
+   registry :- (schema/maybe MetricRegistry)]
+  (let [enabled (ks/to-bool (:enabled config))]
+    (if-not enabled
+      {:profiler nil}
+      (if registry
+        {:profiler (metrics-profiler hostname registry)}
+        (do
```

Comment:
Is it actually possible for this to be false? I don't think it is anymore... it used to be because you could disable metrics globally, but we removed that ability - at some point we started ignoring the `:enabled` setting in the `metrics` section (https://github.com/puppetlabs/trapperkeeper-metrics/blob/0.x/src/puppetlabs/trapperkeeper/services/metrics/metrics_core.clj#L69), and then we removed it all together for (what will be) tk-metrics 1.0.0 (https://github.com/puppetlabs/trapperkeeper-metrics/blob/master/src/clj/puppetlabs/trapperkeeper/services/metrics/metrics_core.clj#L65 - no longer allowed in the config). As far as I can tell, that setting was the only way you would not have a registry at this point.